interchange/DeviceResources.capnp

```diff
@@ -454,6 +477,122 @@ struct Device {
     cellPins @1 : List(CellPinInversion);
   }
 
+  ######################################
+  # Timing modeling
+  #
+  # This section defines the timing model represantation
+  # for the interchange schema.
+  #
+  # Even though there is no strict standard to define how many
+  # corner models need to be defined for a given architecture,
+  # a timing delay model usually includes a "fast" and a
+  # "slow" corner process, each with three delay measures:
+  #  - minimum
+  #  - typical
+  #  - maximum
+  #
+  #  The idea is to have a static definition of the corner models
+  #  so to standardize the interchange format to use at maximum
+  #  two process corner models (fast and slow), with the respective
+  #  delay measures. If an architecture does not include one or two
+  #  (but not all of three) delay measures, the timing model is still valid.
+  #
+  # There are three main location where timing delays must be defined:
+  #  - BEL/Cell pins
+  #  - Wires
+  #  - PIPs
+  #
+  #  * BEL/Cell pins:
+  #  These delays can be sequential (associated to a clock pin) or combinatorial
+  #  (associated to an input/output pair).
+  #  At minimum there are four types of pin delays that need to be considered:
+  #   - comb    : input to output delay
+  #   - setup   : associated to the input and a clock pin
+  #   - hold    : associated to the input and a clock pin
+  #   - clock2q : associated to the clock and the output pin
+  #
+  #  * Wires:
+  #  These delays are associated to wires, or better nodes which are collection
+  #  of electrically connected wires.
+  #  They are described following the RC (Resistance/Capacitance) modeling.
+  #
+  # * PIPs:
+  # These are delays corresponding to the connections between two wires.
+  #
+  ######################################
+
+  # BEL/Cell pins delays
+  struct PinsDelay {
+    firstPin      @0 : PinDelay;
+    secondPin     @1 : PinDelay;
+    cornerModel   @2 : CornerModel;
+    pinsDelayType @3 : PinsDelayType;
+  }
+
+  struct PinDelay {
+    pin   @0 : BELPinIdx $belPinRef();
+    union {
+      noClock   @1 : Void;
+      clockEdge @2 : ClockEdge;
+    }
+  }
+
+  enum ClockEdge {
+    rise @0;
+    fall @1;
+  }
+
+  enum PinsDelayType {
+    comb  @0;
+    setup @1;
+    hold  @2;
+    clk2q @3;
+  }
+
+  # Wire (nodes) delays
+  struct NodeTiming {
+    capacitance @0 : CornerModel;
+    resistance  @1 : CornerModel;
+  }
+
+  # PIP (switches) delays
+  struct PIPTiming {
```

Comment:
We need to define some units for all these values (in particular the delay unit, which could be seconds; nanoseconds or picoseconds - R and C probably only need to be consistent with each other rather than mattering in absolute terms.)

Reply:
I would stick to SI units: ohms, farads, seconds. To get a physically meaningful delay you need to have proper scaling for R and C.

Reply:
My preference would be to choose a unit (picoseconds for delay, for example) that would provide adequate resolution without having to represent fractional numbers.  Integer representation and arithmetic is easier/cheaper and tends to have fewer issues than floating point.